// next/dist/client/next.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderError = exports.router = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var renderError = exports.renderError = function () {
  var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(err) {
    var pathname, query, props;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            pathname = router.pathname, query = router.query;
            _context.next = 3;
            return ErrorComponent.getInitialProps({ err: err, pathname: pathname, query: query });

          case 3:
            props = _context.sent;

            try {
              doRender({ Component: ErrorComponent, props: props });
            } catch (err2) {
              console.error(err2);
            }

          case 5:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function renderError(_x2) {
    return _ref.apply(this, arguments);
  };
}();

exports.render = render;

var _react = require('react');

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _headManager = require('./head-manager');

var _headManager2 = _interopRequireDefault(_headManager);

var _css = require('../lib/css');

var _router = require('../lib/router');

var _app = require('../lib/app');

var _app2 = _interopRequireDefault(_app);

var _evalScript = require('../lib/eval-script');

var _evalScript2 = _interopRequireDefault(_evalScript);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _window = window,
    _window$__NEXT_DATA__ = _window.__NEXT_DATA__,
    component = _window$__NEXT_DATA__.component,
    errorComponent = _window$__NEXT_DATA__.errorComponent,
    props = _window$__NEXT_DATA__.props,
    ids = _window$__NEXT_DATA__.ids,
    err = _window$__NEXT_DATA__.err,
    pathname = _window$__NEXT_DATA__.pathname,
    query = _window$__NEXT_DATA__.query;


var Component = (0, _evalScript2.default)(component).default;
var ErrorComponent = (0, _evalScript2.default)(errorComponent).default;

var router = exports.router = (0, _router.createRouter)(pathname, query, {
  Component: Component,
  ErrorComponent: ErrorComponent,
  ctx: { err: err }
});

var headManager = new _headManager2.default();
var container = document.getElementById('__next');
var defaultProps = { Component: Component, ErrorComponent: ErrorComponent, props: props, router: router, headManager: headManager };

if (ids && ids.length) (0, _css.rehydrate)(ids);

render();

function render() {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  try {
    doRender(props);
  } catch (err) {
    renderError(err);
  }
}

function doRender(props) {
  var appProps = (0, _extends3.default)({}, defaultProps, props);
  _reactDom2.default.render((0, _react.createElement)(_app2.default, appProps), container);
}
