// next/dist/server/build/plugins/watch-pages-plugin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _map = require('babel-runtime/core-js/map');

var _map2 = _interopRequireDefault(_map);

var _path = require('path');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var hotMiddlewareClientPath = (0, _path.join)(__dirname, '..', '..', '..', 'client/webpack-hot-middleware-client');

var defaultPages = new _map2.default();
var _arr = ['_error.js', '_document.js'];
for (var _i = 0; _i < _arr.length; _i++) {
  var p = _arr[_i];
  defaultPages.set((0, _path.join)('bundles', 'pages', p), (0, _path.join)(__dirname, '..', '..', '..', 'pages', p));
}

var WatchPagesPlugin = function () {
  function WatchPagesPlugin(dir) {
    (0, _classCallCheck3.default)(this, WatchPagesPlugin);

    this.dir = (0, _path.resolve)(dir, 'pages');
    this.prevFileDependencies = [];
  }

  (0, _createClass3.default)(WatchPagesPlugin, [{
    key: 'apply',
    value: function apply(compiler) {
      var _this = this;

      compiler.plugin('compilation', function (compilation) {
        compilation.plugin('optimize-assets', function (assets, callback) {
          // transpile pages/_document.js and descendants,
          // but don't need the bundle file
          delete assets[(0, _path.join)('bundles', 'pages', '_document.js')];
          callback();
        });
      });

      compiler.plugin('emit', function (compilation, callback) {
        // watch the pages directory
        compilation.contextDependencies = compilation.contextDependencies.concat([_this.dir]);

        _this.prevFileDependencies = compilation.fileDependencies;

        callback();
      });

      var isPageFile = this.isPageFile.bind(this);
      var getEntryName = function getEntryName(f) {
        return (0, _path.join)('bundles', (0, _path.relative)(compiler.options.context, f));
      };

      compiler.plugin('watch-run', function (watching, callback) {
        (0, _keys2.default)(compiler.fileTimestamps).filter(isPageFile).filter(function (f) {
          return _this.prevFileDependencies.indexOf(f) < 0;
        }).forEach(function (f) {
          var name = getEntryName(f);
          if (defaultPages.has(name)) {
            compiler.removeEntry(name);
          }

          if (compiler.hasEntry(name)) return;

          var entries = [hotMiddlewareClientPath, f + '?entry'];
          compiler.addEntry(entries, name);
        });

        compiler.removedFiles.filter(isPageFile).forEach(function (f) {
          var name = getEntryName(f);
          compiler.removeEntry(name);

          if (defaultPages.has(name)) {
            compiler.addEntry([hotMiddlewareClientPath, defaultPages.get(name) + '?entry'], name);
          }
        });

        callback();
      });
    }
  }, {
    key: 'isPageFile',
    value: function isPageFile(f) {
      return f.indexOf(this.dir) === 0 && (0, _path.extname)(f) === '.js';
    }
  }]);
  return WatchPagesPlugin;
}();

exports.default = WatchPagesPlugin;
