// next/dist/server/build/webpack.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _map = require('babel-runtime/core-js/map');

var _map2 = _interopRequireDefault(_map);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _path = require('path');

var _crypto = require('crypto');

var _fs = require('fs');

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _globPromise = require('glob-promise');

var _globPromise2 = _interopRequireDefault(_globPromise);

var _writeFileWebpackPlugin = require('write-file-webpack-plugin');

var _writeFileWebpackPlugin2 = _interopRequireDefault(_writeFileWebpackPlugin);

var _friendlyErrorsWebpackPlugin = require('friendly-errors-webpack-plugin');

var _friendlyErrorsWebpackPlugin2 = _interopRequireDefault(_friendlyErrorsWebpackPlugin);

var _unlinkFilePlugin = require('./plugins/unlink-file-plugin');

var _unlinkFilePlugin2 = _interopRequireDefault(_unlinkFilePlugin);

var _watchPagesPlugin = require('./plugins/watch-pages-plugin');

var _watchPagesPlugin2 = _interopRequireDefault(_watchPagesPlugin);

var _watchRemoveEventPlugin = require('./plugins/watch-remove-event-plugin');

var _watchRemoveEventPlugin2 = _interopRequireDefault(_watchRemoveEventPlugin);

var _dynamicEntryPlugin = require('./plugins/dynamic-entry-plugin');

var _dynamicEntryPlugin2 = _interopRequireDefault(_dynamicEntryPlugin);

var _detachPlugin = require('./plugins/detach-plugin');

var _detachPlugin2 = _interopRequireDefault(_detachPlugin);

var _jsonPagesPlugin = require('./plugins/json-pages-plugin');

var _jsonPagesPlugin2 = _interopRequireDefault(_jsonPagesPlugin);

var _config = require('../config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var documentPage = (0, _path.join)('pages', '_document.js');
var defaultPages = ['_error.js', '_error-debug.js', '_document.js'];

exports.default = function () {
  var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(dir) {
    var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        _ref2$dev = _ref2.dev,
        dev = _ref2$dev === undefined ? false : _ref2$dev,
        _ref2$quiet = _ref2.quiet,
        quiet = _ref2$quiet === undefined ? false : _ref2$quiet;

    var config, pages, entry, defaultEntries, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, p, nextPagesDir, interpolateNames, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _p, entryName, path, nextNodeModulesDir, minChunks, plugins, mainBabelOptions, hasBabelRc, rules, webpackConfig;

    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            dir = (0, _path.resolve)(dir);
            config = (0, _config2.default)(dir);
            _context.next = 4;
            return (0, _globPromise2.default)('pages/**/*.js', { cwd: dir });

          case 4:
            pages = _context.sent;
            entry = {
              'main.js': dev ? require.resolve('../../client/next-dev') : require.resolve('../../client/next')
            };
            defaultEntries = dev ? [(0, _path.join)(__dirname, '..', '..', 'client/webpack-hot-middleware-client')] : [];
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            _context.prev = 10;

            for (_iterator = (0, _getIterator3.default)(pages); !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
              p = _step.value;

              entry[(0, _path.join)('bundles', p)] = defaultEntries.concat(['./' + p + '?entry']);
            }

            _context.next = 18;
            break;

          case 14:
            _context.prev = 14;
            _context.t0 = _context['catch'](10);
            _didIteratorError = true;
            _iteratorError = _context.t0;

          case 18:
            _context.prev = 18;
            _context.prev = 19;

            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }

          case 21:
            _context.prev = 21;

            if (!_didIteratorError) {
              _context.next = 24;
              break;
            }

            throw _iteratorError;

          case 24:
            return _context.finish(21);

          case 25:
            return _context.finish(18);

          case 26:
            nextPagesDir = (0, _path.join)(__dirname, '..', '..', 'pages');
            interpolateNames = new _map2.default();
            _iteratorNormalCompletion2 = true;
            _didIteratorError2 = false;
            _iteratorError2 = undefined;
            _context.prev = 31;


            for (_iterator2 = (0, _getIterator3.default)(defaultPages); !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
              _p = _step2.value;
              entryName = (0, _path.join)('bundles', 'pages', _p);
              path = (0, _path.join)(nextPagesDir, _p);

              if (!entry[entryName]) {
                entry[entryName] = defaultEntries.concat([path + '?entry']);
              }
              interpolateNames.set(path, 'dist/pages/' + _p);
            }

            _context.next = 39;
            break;

          case 35:
            _context.prev = 35;
            _context.t1 = _context['catch'](31);
            _didIteratorError2 = true;
            _iteratorError2 = _context.t1;

          case 39:
            _context.prev = 39;
            _context.prev = 40;

            if (!_iteratorNormalCompletion2 && _iterator2.return) {
              _iterator2.return();
            }

          case 42:
            _context.prev = 42;

            if (!_didIteratorError2) {
              _context.next = 45;
              break;
            }

            throw _iteratorError2;

          case 45:
            return _context.finish(42);

          case 46:
            return _context.finish(39);

          case 47:
            nextNodeModulesDir = (0, _path.join)(__dirname, '..', '..', '..', 'node_modules');
            minChunks = pages.filter(function (p) {
              return p !== documentPage;
            }).length;
            plugins = [new _webpack2.default.LoaderOptionsPlugin({
              options: {
                context: dir,
                customInterpolateName: function customInterpolateName(url, name, opts) {
                  return interpolateNames.get(this.resourcePath) || url;
                }
              }
            }), new _writeFileWebpackPlugin2.default({
              exitOnErrors: false,
              log: false,
              // required not to cache removed files
              useHashIndex: false
            }), new _webpack2.default.optimize.CommonsChunkPlugin({
              name: 'commons',
              filename: 'commons.js',
              minChunks: Math.max(2, minChunks)
            }), new _jsonPagesPlugin2.default()];


            if (dev) {
              plugins.push(new _webpack2.default.HotModuleReplacementPlugin(), new _webpack2.default.NoErrorsPlugin(), new _detachPlugin2.default(), new _dynamicEntryPlugin2.default(), new _unlinkFilePlugin2.default(), new _watchRemoveEventPlugin2.default(), new _watchPagesPlugin2.default(dir));
              if (!quiet) {
                plugins.push(new _friendlyErrorsWebpackPlugin2.default());
              }
            } else {
              plugins.push(new _webpack2.default.DefinePlugin({
                'process.env.NODE_ENV': (0, _stringify2.default)('production')
              }), new _webpack2.default.optimize.UglifyJsPlugin({
                compress: { warnings: false },
                sourceMap: false
              }));
            }

            mainBabelOptions = {
              babelrc: true,
              cacheDirectory: true,
              sourceMaps: dev ? 'both' : false,
              presets: []
            };
            hasBabelRc = (0, _fs.existsSync)((0, _path.join)(dir, '.babelrc'));

            if (hasBabelRc) {
              console.log('> Using .babelrc defined in your app root');
            } else {
              mainBabelOptions.presets.push(require.resolve('./babel/preset'));
            }

            rules = (dev ? [{
              test: /\.js(\?[^?]*)?$/,
              loader: 'hot-self-accept-loader',
              include: [(0, _path.join)(dir, 'pages'), nextPagesDir]
            }, {
              test: /\.js(\?[^?]*)?$/,
              loader: 'react-hot-loader/webpack',
              exclude: /node_modules/
            }] : []).concat([{
              test: /\.json$/,
              loader: 'json-loader'
            }, {
              test: /\.(js|json)(\?[^?]*)?$/,
              loader: 'emit-file-loader',
              include: [dir, nextPagesDir],
              exclude: function exclude(str) {
                return (/node_modules/.test(str) && str.indexOf(nextPagesDir) !== 0
                );
              },

              options: {
                name: 'dist/[path][name].[ext]'
              }
            }, {
              loader: 'babel-loader',
              include: nextPagesDir,
              options: {
                babelrc: false,
                cacheDirectory: true,
                sourceMaps: dev ? 'both' : false,
                plugins: [[require.resolve('babel-plugin-module-resolver'), {
                  alias: {
                    'ansi-html': require.resolve('ansi-html'),
                    'styled-jsx/style': require.resolve('styled-jsx/style')
                  }
                }]]
              }
            }, {
              test: /\.js(\?[^?]*)?$/,
              loader: 'babel-loader',
              include: [dir, nextPagesDir],
              exclude: function exclude(str) {
                return (/node_modules/.test(str) && str.indexOf(nextPagesDir) !== 0
                );
              },

              query: mainBabelOptions
            }]);
            webpackConfig = {
              context: dir,
              entry: entry,
              output: {
                path: (0, _path.join)(dir, '.next'),
                filename: '[name]',
                libraryTarget: 'commonjs2',
                publicPath: '/_webpack/',
                strictModuleExceptionHandling: true,
                devtoolModuleFilenameTemplate: function devtoolModuleFilenameTemplate(_ref3) {
                  var resourcePath = _ref3.resourcePath;

                  var hash = (0, _crypto.createHash)('sha1');
                  hash.update(Date.now() + '');
                  var id = hash.digest('hex').slice(0, 7);

                  // append hash id for cache busting
                  return 'webpack:///' + resourcePath + '?' + id;
                }
              },
              resolve: {
                modules: [nextNodeModulesDir, 'node_modules'].concat((process.env.NODE_PATH || '').split(process.platform === 'win32' ? ';' : ':').filter(function (p) {
                  return !!p;
                }))
              },
              resolveLoader: {
                modules: [nextNodeModulesDir, 'node_modules', (0, _path.join)(__dirname, 'loaders')]
              },
              plugins: plugins,
              module: {
                rules: rules
              },
              devtool: dev ? 'inline-source-map' : false,
              performance: { hints: false }
            };

            if (!config.webpack) {
              _context.next = 61;
              break;
            }

            console.log('> Using "webpack" config function defined in next.config.js.');
            _context.next = 60;
            return config.webpack(webpackConfig, { dev: dev });

          case 60:
            webpackConfig = _context.sent;

          case 61:
            return _context.abrupt('return', (0, _webpack2.default)(webpackConfig));

          case 62:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[10, 14, 18, 26], [19,, 21, 25], [31, 35, 39, 47], [40,, 42, 46]]);
  }));

  function createCompiler(_x) {
    return _ref.apply(this, arguments);
  }

  return createCompiler;
}();
